crates/matrix-sdk/src/widget_api/handler/incoming.rs

```diff
@@ -1,19 +1,12 @@
 use serde::{Deserialize, Serialize};
-
 use url::Url;
 
-use crate::widget_api::messages::message::MessageBody;
-
-use super::Request;
+pub use super::{Error, Request};
 
-#[derive(Serialize, Deserialize, Debug, Clone)]
-#[serde(tag = "action")]
-pub enum FromWidgetMessage {
-    #[serde(rename = "supported_api_versions")]
+#[allow(missing_debug_implementations)]
+pub enum Message {
     GetSupportedApiVersion(Request<(), SupportedVersions>),
-    #[serde(rename = "content_loaded")]
     ContentLoaded(Request<(), ()>),
-    #[serde(rename = "org.matrix.msc2931.navigate")]
     Navigate(Request<Url, Result<(), &'static str>>),
 }
```

Comment:
Hey Timo, good question, let me reply!

So the `FromWidget` from the `messages` module is an enum that represents **the content of JSON messages** that we get from the widget. The `Message` from `incoming` module represents the **incoming requests** from the widget with the reply object wrapped into them so that the request's handler can send the reply to the sender.

In other words, `FromWidget` is a set of all messages we can get from the widget (including the request bodies, and the widget's responses to our requests). It's something that we use for serialization/deserialization of the transport implementation (i.e. a rather "low-level" thing). Whereas the `incoming/Message` is the set of types that represent requests along with their responses (i.e. it's a subset of `FromWidget` in the sense that it only contains incoming requests and it's a superset of a `FromWidget` in a sense that it additionally contains the `Request` logic, i.e. the `Reply` system that the `FromWidget` does not have). That's why they have different types.

Does it make any sense? 🙂

Reply:
Yes it definitly makes sense how they are different. Maybe there still is a solution to unify them. The two enums Incoming and From Widget will have the same enum cases and each enum case uses the same combination of request and resposne body types. It only differs, that one is using Message and the other Request. And message and request also only differ that the message contains the header and the request the reply.
Maybe it is possible to add an Option<Reply> to the Message. Than we can populate the message with the reply in the transport and inside the handler the message can be used like the request?

Reply:
> Yes it definitly makes sense how they are different. Maybe there still is a solution to unify them.

Well, they represent different things. I'm not sure unifying them is desired even if it was possible.

> The two enums Incoming and From Widget will have the same enum cases and each enum case uses the same combination of request and resposne body types.

Not exactly so, the `FromWidget` will have more messages as it defines all messages that can be received from a widget. This includes the requests, but also the types of replies. I.e. the `FromWidget` is the serialize/deserialize part of the protocol. Whereas the `Incoming` is a part that is tied to the "business logic" (API request/response semantics). `FromWidget` has more data (including low-level transport fields, and the replies). `Incoming` has less data, but more functionality (request/reply) logic. Also they don't map 1:1, e.g. 2 objects of type `FromWidget` do correspond to 1 object of the `Incoming`.

> Maybe it is possible to add an Option to the Message. Than we can populate the message with the reply in the transport and inside the handler the message can be used like the request?

I'm not quite sure how it would look like 🙂